### PR TITLE
SC-647-casetype-destruction

### PR DESCRIPTION
**Pull request by potan**
Impruve match/case
1. Use constants
2. Dectruction caseType

### CODE REVIEW

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala [hunk 1]
```diff
@@ -238,36 +237,40 @@ object ExpressionCompiler {
...
-      exprTypesWithErr <- (typedExpr.t match {
-        case u: UNION => u.pure[CompileM]
-        case _        => raiseError[Id, CompilerContext, CompilationError, UNION](MatchOnlyUnion(p.start, p.end))
-      }).handleError()
-      exprTypes = exprTypesWithErr._1.getOrElse(NOTHING)
+//      exprTypesWithErr <- (typedExpr.t match {
+//        case u: UNION => u.pure[CompileM]
+//        case _        => raiseError[Id, CompilerContext, CompilationError, UNION](MatchOnlyUnion(p.start, p.end))
+//      }).handleError()
+      exprTypes = typedExpr.t //exprTypesWithErr._1.getOrElse(NOTHING)
```
Comment (finnoio) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala:252]: remove comments

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala [hunk 2]
```diff
@@ -277,18 +280,22 @@ object ExpressionCompiler {
...
+              (exprTypes equivalent matchedTypesUnion)),
               (),
               MatchNotExhaustive(p.start, p.end, exprTypes.typeList, matchTypes)
             )
             .toCompileM
             .handleError()
       _ <- set[Id, CompilerContext, CompilationError](ctx.copy(tmpArgsIdx = tmpArgId))
 
-      errorList = exprTypesWithErr._2 ++ ifCasesWithErr._2 ++ compiledMatch.errors ++ checkWithErr._2
+      errorList = /*exprTypesWithErr._2 ++*/ ifCasesWithErr._2 ++ compiledMatch.errors ++ checkWithErr._2
```
Comment (finnoio) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala:298]: remove comment

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala [hunk 3]
```diff
@@ -639,55 +646,101 @@ object ExpressionCompiler {
...
+          mc.expr
+        } else {
+          mc.pattern.fields.foldLeft(mc.expr) {
+            (e, v) => v match {
+              case Expressions.PBind(Some(v@Expressions.PART.VALID(pos, name)), Expressions.PART.VALID(_, a)) =>
+                val ty = caseType.fields.find(_._1 == a).map(_._2)
+                /*match {
+                  case Some((_,ty)) => ty
+                  case _ => ???
+                }*/
```
Comment (finnoio) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala:676]: remove commented lines

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Expressions.scala [hunk 4]
```diff
@@ -137,10 +137,19 @@ object Expressions {
...
+  case class PConst(consts: Seq[EXPR], override val accessor: PART[String]) extends PVal
+
+  case class Pattern(name: Option[PART[String]], ty: Option[Type], fields: Seq[PVal], consts: Seq[EXPR])
+
   case class MATCH_CASE(
       position: Pos,
-      newVarName: Option[PART[String]],
-      caseType: Type,
+      //newVarName: Option[PART[String]],
+      //caseType: Type,
```
Comment (finnoio) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Expressions.scala:151]: remove commented lines

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Parser.scala [hunk 5]
```diff
@@ -272,28 +272,89 @@ object Parser {
...
+            error => MATCH_CASE(error.position, pattern = p, expr = error),
             { pos =>
-              val exprStart = pos.orElse(v.map(_.position)).fold(caseStart)(_.end)
+              val exprStart = pos.orElse(p.name.map(_.position)).fold(caseStart)(_.end)
               MATCH_CASE(
                 Pos(caseStart, end),
-                newVarName = v,
-                caseType = types,
+                //newVarName = v,
+                //caseType = types,
```
Comment (finnoio) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Parser.scala:356]: remove commented lines

**Review by finnoio — CHANGES_REQUESTED** [review]

**Review by xrtm000 — CHANGES_REQUESTED** [review]
Seems strange that works this code
```
match (if true then "string" else 1) {
  case "string" => "yes"
  case _        => "no"
}
```
and such one
```
match (if true then "string" else 1) {
  case 1 => "yes"
  case _ => "no"
}
```
but if we define both possible values
```
match (if true then "string" else 1) {
  case "string" => "yes"
  case 1        => "no"
}
```
compilation will be failed with enigmatic error: _"Matching not exhaustive: **possibleTypes are String, Int, while matched are String, Int**"_

**Comment by xrtm000** [comment]
Should we support mix of destructions? 
```
match (5, PointC(1)) {
  case (5, PointC(YB=1)) => "yes"
  case _                 => "no"
}
```
```
match PointC(1) {
  case PointC(YB=1) | PointC(YB=2) => "yes"
  case _                           => "no"
}
```
It might be useful, but currently it is not supported.

File: lang/tests/src/test/scala/com/wavesplatform/lang/ScriptParserTest.scala [hunk 6]
```diff
@@ -782,7 +782,7 @@ class ScriptParserTest extends PropSpec with PropertyChecks with Matchers with S
     )
   }
 
-  property("pattern matching with invalid case - expression in variable definition") {
+  ignore("pattern matching with invalid case - expression in variable definition") {
```
Comment (xrtm000) [inline lang/tests/src/test/scala/com/wavesplatform/lang/ScriptParserTest.scala:785]: Please, adapt this test

**Review by xrtm000 — CHANGES_REQUESTED** [review]

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Parser.scala [hunk 7]
```diff
@@ -272,28 +272,89 @@ object Parser {
...
+              case (_, Var(_, _), _) => true
+              case _ => false
+            }
+            val pats = p.zipWithIndex.map {
+              case ((s, Const(c), e), i) => PConst(c, PART.VALID(Pos(s,e), s"_${i+1}"))
+              case ((s, Var(REF(_, v, _, _), ty), e), i) => PBind(Some(v), PART.VALID(Pos(s,e), s"_${i+1}"))
+            }
+            (allTyped, allUntyped) match {
+              case (true, false) => Pattern(None, Some(Tuple(types)), pats, Seq())
+              case (false, false) => Pattern(Some(PART.INVALID(Pos(start, end), "destructed tuple is not fully typed")), Some(Tuple(Seq())), pats, Seq())
```
Comment (xrtm000) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Parser.scala:330]: Please add test that rises "destructed tuple is not fully typed" error

**Comment by mrkraft** [comment]
Unfinished

**Pull request closed without merging**